[Core] Reduce TTFT with concurrent partial prefills

joerunde opened the pull request:
Replaces #10061, as inspired by @njhill and @comaniac's comments. Co-authored by @prashantgupta24 

Context: our customers running large multi-tenanted SaaS deployments of vLLM have a problem where high volumes of small-prompt requests are usually processed smoothly, but quickly pile up in a giant queue when a small number of large-prompt requests are submitted. We see the decoding throughput drop to zero on multiple replicas when this happens.

The current chunked prefill implementation only allows a single sequence to be partially prefilled at a time. This has a few limitations:
- Multiple medium-sized prompts must wait to be prefilled serially, increasing TTFT for those in the back of the queue
- A single very large prompt will block all other prompts from prefilling for many iterations. This can eventually starve decoding- for example a 130k token prompt with `—max-num-batched-tokens=512` will take about 250 iterations to prefill, in which time the currently decoding sequences may all finish. Send a few of these requests at once and very quickly nothing will be decoding.

This PR implements both
- An explicit setting for the number of sequences that can be partially prefilled concurrently. This can be configured with `--max-num-partial-prefills=N`
- A limit on the number of “very long prompt” sequences that can be prefilled concurrently. This can be configured with
  - `--max-long-partial-prefills=N` to set the limit on the number of long sequences that can be concurrently prefilled. This defaults to 1 sequence.
  - `--long-prefill-threashold=x%` to set a percentage of the context length that determines which sequences are considered "long". This defaults to 4%

This is implemented in the v0 scheduler. We’re aware that the v1 implementation is underway and will later become the default, but we need a fix for our customers soon and we hope that what we discover here may help inform a different, better solution in the v1 scheduler.

To test this we created three scenarios, a “medium request” case, a “large request” case, and a “mixed” case.

For the medium request case, we created a subset of the `sharegpt` dataset with 900 small requests (<50 prompt characters) and 100 of the largest requests (typically between 10k and 20k prompt characters, which we call “medium” sized). We modified the `benchmark_serving.py` test to not filter out any of the small or large requests, and ran it with this dataset. What we expect to find is similar throughput compared to the main branch, but much lower TTFT on the small requests. Since 10% of the requests are larger than the rest, we should see better TTFT at p90 and below, with comparable TTFT above p90.

For the large request case, we took 990 of the smallest requests from the `sharegpt` dataset, and then took 10 of the largest requests and duplicated the prompts until they were around 100k characters in length. We ran this in the same way as the medium request case, and here we expect to see smaller TTFT across the board since the small requests will no longer be blocked from prefilling by the few very large requests.

For the mixed case, we used 850 “small”, and 140 “medium” requests, as well as 10 "large" requests where we duplicated the prompts up to 200k characters.

All tests were run on a single 80GB A100, with the command:
```
python benchmarks/benchmark_serving.py --model meta-llama/Meta-Llama-3.1-8B-Instruct --dataset-path ${test_case} --metric-percentiles 80,85,90,95,99 --request-rate 12
```

We ran the tests against the main branch (commit `874f551b3626321f6bf9a902b8fd9fc1fa7c7f2e`), as well as this PR with the new optimization both disabled (`--max-num-partial-prefills=1`), and enabled (`--max-num-partial-prefills=4`)

The results are shown here:
![results](https://github.com/user-attachments/assets/26782cb2-16e2-4528-9c3e-ee833e1ba7ea)

The TTFT improvements are very easy to see- in the medium case we cut the p90 TTFT in half, and in the large case we cut it nearly 30x. In both cases we did not measure a throughput drop when run with  `--max-num-partial-prefills=1`, and the throughput drop with `--max-num-partial-prefills=4` is minimal.

Surprisingly, along with the massive TTFT improvements in the "mixed" test case, we also see a 4% throughput _improvement_ (3506 tokens/s up from 3368 tokens/s). Based on the fact that ITL still looks a little slower, it seems that the throughput is higher simply because more requests were able to be successfully scheduled at the same time.

cc @rickyyx 

---

<details>
<!-- inside this <details> section, markdown rendering does not work, so we use raw html here. -->
<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>


</details>

joerunde committed:
:bug: fix multi-chunked-prefill sampler bug

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

prashantgupta24 committed:
🚧 add num_prefill_slots arg

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

joerunde committed:
:sparkles: start to write prefill slot logic

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

prashantgupta24 committed:
🎨 format

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

joerunde committed:
:sparkles: update num tokens for prefill slots

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

prashantgupta24 committed:
♻️ add schedule_chunked_prefill logic

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

prashantgupta24 committed:
♻️ change function name

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

joerunde committed:
:sparkles: reserve incoming prefill slots

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

prashantgupta24 committed:
🎨 fix some typos

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

joerunde committed:
:zap: finish awesome scheduler

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

joerunde committed:
:bug: fix the deadlocks

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

joerunde committed:
:memo: Add more docstrings

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

joerunde committed:
:bug: fix deadlock

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

joerunde commented on vllm/model_executor/layers/sampler.py:
```diff
@@ -1262,6 +1264,10 @@ def _build_sampler_output(
+        assert len(sampling_metadata.seq_groups) \
+            == len(maybe_deferred_sample_results) \
+            == len(prompt_logprobs) \
+            == len(sample_logprobs)
```
I'm not really familiar with the sampler code at all, but this assert at least does not trigger under any of our tests so far.

@comaniac @rickyyx Do y'all have pointers to any other places we should dig into where there might be an implicit assumption that the number of partial prefills is <= 1?

rickyyx replied:
I think the assertions currently would only happen if:
- All running sequences are chunked continuous prefills
- No decode sequences in the scheduler batch.

joerunde replied:
Sweet, we will try to add some unit tests to cover that case and see if we can make it crash. Thanks!

joerunde commented on vllm/core/scheduler.py:
```diff
@@ -395,6 +395,38 @@ def __init__(
+        # requests may jump the queue in front of them and get to the decode
+        # phase faster.
+        self.big_prefill_threshold = scheduler_config.max_model_len // 25
+        self.max_big_requests = 1  # TODO: something
```
We think that leaving this set to `1` is reasonable, since in the worst case when there are many large prompt requests this is identical to the existing behavior. We could add a new cli arg or environment variable to adjust it, but we also don't really want to pollute the config surface even more.

joerunde replied:
Well they could always turn off the feature entirely :D

But yeah you're right. Do you mind if this is a` vllm.envs` setting instead of another cli arg?

joerunde committed:
:construction: WIP scheduler tests

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

joerunde committed:
:bug: fix prefix caching

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

joerunde committed:
:test_tube: add prefix caching test

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

prashantgupta24 commented on vllm/core/scheduler.py:
```diff
@@ -1124,6 +1184,19 @@ def _schedule_chunked_prefill(self) -> SchedulerOutputs:
+
+        # Set slot counts for next iteration
+        self.prefill_slots_running = len(prefilling)
+        self.big_prefill_requests = len([
```
I'm wondering if this block of code will handle any cases that are not covered by the increment we do within `_schedule_prefills` 

https://github.com/vllm-project/vllm/pull/10235/files#diff-0e64e07127ff20fc01be18cc95c0f58e0ff818e0dbabd5e8d2e07ca27943a7e4R1001-R1002

joerunde replied:
Ah, yeah good catch. We should add a second iteration on the unit test for large requests to make sure that this code block counts the mid-prefill large request and the second one doesn't immediately start. Would you mind adding that?

prashantgupta24 replied:
Done

prashantgupta24 committed:
✅ add second test iteration

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

prashantgupta24 commented on vllm/core/scheduler.py:
```diff
@@ -882,6 +922,17 @@ def _schedule_prefills(
+                waiting_queue.popleft()
+                continue
+
             num_new_tokens = self._get_num_new_tokens(seq_group,
                                                       SequenceStatus.WAITING,
                                                       enable_chunking, budget)
```
In the case where we only have 1 `prefill-slot` and we send in 2 concurrent requests, one big and one small, when the scheduler is trying to `schedule_prefill`, it fills up the token budget with the first large request. When it loops through the waiting requests and tries to get `num_new_tokens` for the 2nd request, it gets 0 since budget is 0. Can we break out earlier instead of breaking out way down on line 990?

joerunde replied:
Yeah that sounds reasonable. Maybe at the very end of the loop after we schedule a request, we can check if budget == 0 and `break` instead of looping to the next request?

Could be ⚡⚡⚡ 🚀

joerunde replied:
I think in cases where you have 1 prefill slot and a large request running, most of the time it'll use up the budget inside of `_schedule_running` and hit the check at the top of `_schedule_prefills` and not loop at all, but this optimization would still be useful. In cases with constant load of small requests, you could potentially save one loop iteration most times here

prashantgupta24 committed:
✅ add llm engine test

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

prashantgupta24 committed:
♻️ quicker budget check

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

comaniac commented on vllm/core/scheduler.py:
```diff
@@ -935,18 +993,15 @@ def _schedule_prefills(
             waiting_queue.popleft()
             self._allocate_and_set_running(seq_group)
 
+            if is_big:
+                self.big_prefill_requests += 1
```
It seems improper to put `big_prefill_requests` as a scheduler attribute, because its life cycle is jus per scheduling step. If you need to pass this value to various functions when scheduling one step, you should just pass it as needed.

joerunde replied:
As is the lifecycle is across scheduling steps, we use the values from the previous iteration for the next one. 

We could probably change it to only live per-scheduling-step though, and encapsulate all these values into a `PartialPrefillMetadata` to pass around. Would that be better?

joerunde replied:
Updated to calculate the required metadata during each iteration instead of saving it across iterations. We didn't measure any significant performance drop from doing that

joerunde commented on vllm/config.py:
```diff
@@ -1085,6 +1085,7 @@ def __init__(self,
                  max_num_batched_tokens: Optional[int],
                  max_num_seqs: int,
                  max_model_len: int,
+                 num_prefill_slots: int = 1,
```
It's technically only the number of partial prefills allowed in a batch. You could still have like 100 sequence groups with 5 prompt tokens each all schedule in a single step here.

`max_num_partial_prefills`?

joerunde commented on vllm/core/scheduler.py:
```diff
@@ -395,6 +395,38 @@ def __init__(
+        # The number of big prefill requests is limited so that smaller
+        # requests may jump the queue in front of them and get to the decode
+        # phase faster.
+        self.big_prefill_threshold = scheduler_config.max_model_len // 25
```
Yeah, `long` is way better!

joerunde commented on vllm/core/scheduler.py:
```diff
@@ -395,6 +395,38 @@ def __init__(
+        self.prefill_chunk_sizes = defaultdict(
+            lambda: scheduler_config.max_num_batched_tokens // self.
+            num_prefill_slots)
+        self.prefill_chunk_sizes[0] = scheduler_config.max_num_batched_tokens
+        for i in range(1, self.num_prefill_slots):
+            self.prefill_chunk_sizes[i] = \
+                scheduler_config.max_num_batched_tokens // i
```
The reasoning is in the docstring- it's a defaultdict so that when there are potentially many prefills running at the same time we cap the minimum per-slot budget at `budget // self.num_prefill_slots`

We can use a list instead though and try to ensure that we never access past the end

joerunde commented on vllm/core/scheduler.py:
```diff
@@ -882,9 +922,27 @@ def _schedule_prefills(
+            is_big = self._is_big_seq_group(seq_group)
+            if is_big and self.big_prefill_requests >= self.max_big_requests \
+                and self.num_prefill_slots > 1:
+                # When we have more than one prefill slot, we limit the number
```
I think the subject-verb agreement is correct as is? But we'll probably change this naming anyway re: your previous comments

joerunde commented on vllm/core/scheduler.py:
```diff
@@ -395,6 +395,38 @@ def __init__(
+        self.num_prefill_slots = scheduler_config.num_prefill_slots
+        self.prefill_slots_running = 0
+        self.big_prefill_requests = 0
+        # Requests with more than (4% max context length) tokens to prefill
+        # are "big".
```
The entire goal here is to not allow decode to be starved by the prefill phase blocking on long requests- this part of the PR description:
> A single very large prompt will block all other prompts from prefilling for many iterations. This can eventually starve decoding- for example a 130k token prompt with —max-num-batched-tokens=512 will take about 250 iterations to prefill, in which time the currently decoding sequences may all finish. Send a few of these requests at once and very quickly nothing will be decoding.

Just allowing concurrent partial prefills doesn't solve the problem by itself, because multiple long requests could still block up the prefill. So what we do is only allow a single long request to prefill, and allow smaller requests to be pulled from the waiting queue instead of more long ones

prashantgupta24 committed:
🎨 rename to max_num_partial_prefills

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

prashantgupta24 committed:
🎨 more renaming to max_num_partial_prefills + docstring updates

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

prashantgupta24 committed:
🎨 rename big to long

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

prashantgupta24 committed:
♻️ add cli args for partial_prefill configs

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

prashantgupta24 committed:
🎨 fix request word typo

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

prashantgupta24 committed:
🎨 more docstring changes

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

prashantgupta24 committed:
🎨 forgot to add the new args to config

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

prashantgupta24 committed:
🐛 fix range bug on partial_prefill_budget_lookup_list

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

prashantgupta24 committed:
🎨 add docstring to test function

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

joerunde committed:
:construction: WIP move metadata to dataclass

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

prashantgupta24 committed:
🎨 wrap up PartialPrefillMetadata

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

prashantgupta24 committed:
♻️ add some utility functions within partial_prefill_metadata

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

prashantgupta24 committed:
🎨 change to long_prefill_token_threshold

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

prashantgupta24 commented on vllm/core/scheduler.py:
```diff
@@ -935,18 +1092,15 @@ def _schedule_prefills(
-                    or not budget.can_schedule(num_new_tokens=num_new_tokens,
-                                               num_new_seqs=num_new_seqs)):
-                break
-
```
This block of code was just moved up

prashantgupta24 committed:
🔥 remove commented code

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

prashantgupta24 committed:
🐛 fix the big bug! (Thanks Joe)

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

joerunde committed:
:memo: docstings galore

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

prashantgupta24 commented on vllm/core/scheduler.py:
```diff
@@ -295,6 +295,96 @@ def scheduled_seq_group_builder():
+                    and self.scheduler_config.max_num_partial_prefills > 1
+
+    def increment_partial_prefills(self, seq_group: SequenceGroup):
+        # When a new prefill is scheduled, we need to know if it is a
```
```suggestion
        # When a new prefill is scheduled, we need to know if it is a
        # long request
```

prashantgupta24 committed:
🎨 fix typo

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

prashantgupta24 committed:
⏪ revert logging change

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

joerunde commented on tests/core/test_chunked_prefill_scheduler.py:
```diff
@@ -507,3 +646,86 @@ def test_perfix_caching():
+
+@pytest.mark.parametrize("model", ["facebook/opt-125m"])
+@pytest.mark.parametrize("max_num_partial_prefills", [2, 4, 8])
+def test_chunked_prefill_with_actual_engine(model: str,
```
cc @rickyyx here's what we tried to do to test that the sampler doesn't throw any assertions- we put multiple prompts into an engine and manually step it forward with them all partially prefilled

prashantgupta24 committed:
✅ remove value error from test

not sure if that's needed?

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

prashantgupta24 committed:
✅ remove value error from test

not sure if that's needed?

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

prashantgupta24 committed:
🎨 fix typo

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

prashantgupta24 committed:
✅ make test comprehensive

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

prashantgupta24 commented on tests/core/test_chunked_prefill_scheduler.py:
```diff
@@ -121,6 +124,210 @@ def test_chunk():
+
+    # in the fourth iteration, both small requests are decoding
+    # so large request gets all the budget
+    seq_group_meta, out = schedule_and_update_computed_tokens(scheduler)
+    # large req gets 63 tokens (minus 1 for decode)
+    assert seq_group_meta[0].token_chunk_size == 63
+    assert seq_group_meta[1].token_chunk_size == 1  # decode
```
Not sure if this is a bug, but at this stage, request-3 should be decoding, but it didn't get any budget. Request-2 got budget for `1` decode token, and request-0 got the remaining budget for prefilling `63` tokens. Is that expected? 

Based on this comment,
```
        # Update new running requests.
        # By default, vLLM scheduler prioritizes prefills.
        # Once chunked prefill is enabled,
        # the policy is changed to prioritize decode requests.
```

vllm should have prioritized decode requests and given request-3 `1` budget already, and request-0 `62`?

prashantgupta24 replied:
This does happen in the next iteration though, but not sure why it didn't happen in this iteration

joerunde replied:
Good catch- I finally went back and looked at this. It's not really much of an issue since the request skips at most 1 step of decode, but I found the bug where if a partial prefill is ordered before a full prefill in the `self.running` queue then the partial prefill is prioritized over the new decode in the next step. Fixed that up 👍

prashantgupta24 committed:
🎨 fix unused vars in test

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

prashantgupta24 committed:
🎨 some more comments

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

prashantgupta24 committed:
🎨 fix merge conflict

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

prashantgupta24 committed:
🎨 fmt

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

njhill commented:
@prashantgupta24 it looks like the regression test is hanging for some reason, after prompt processing starts.
```
[2024-11-21T18:53:40Z] Running 4 items in this shard: tests/test_regression.py::test_duplicated_ignored_sequence_group, tests/test_regression.py::test_max_tokens_none, tests/test_regression.py::test_gc, tests/test_regression.py::test_model_from_modelscope
...
[2024-11-21T18:53:51Z] INFO 11-21 10:53:51 model_runner.py:1399] Capturing cudagraphs for decoding. This may lead to unexpected consequences if the model is not static. To run the model in eager mode, set 'enforce_eager=True' or use '--enforce-eager' in the CLI.
[2024-11-21T18:53:51Z] INFO 11-21 10:53:51 model_runner.py:1403] If out-of-memory error occurs during cudagraph capture, consider decreasing `gpu_memory_utilization` or switching to eager mode. You can also reduce the `max_num_seqs` as needed to decrease memory usage.
[2024-11-21T18:54:02Z] INFO 11-21 10:54:02 model_runner.py:1517] Graph capturing finished in 10 secs, took 0.12 GiB
Processed prompts:  50% 1/2 [00:00<00:00,  1.70it/s, est. speed input: 10.17 toks/s, output: 434.01 toks/s]
# Received cancellation signal, interrupting
[2024-11-21T20:52:13Z] 🚨 Error: The command exited with status -1
```

prashantgupta24 committed:
♻️ merge with main

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

prashantgupta24 committed:
Merge branch 'main' into prefill-slots

prashantgupta24 committed:
🎨 fix fmt

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

prashantgupta24 committed:
⏪ revert quick budget check

Turns out we need to add seqs to SequenceStatus.FINISHED_IGNORED
if num_new_tokens > prompt_limit

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

prashantgupta24 committed:
🎨 fmt

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

prashantgupta24 commented:
> @prashantgupta24 it looks like the regression test is hanging for some reason, after prompt processing starts.
> 
> ```
> [2024-11-21T18:53:40Z] Running 4 items in this shard: tests/test_regression.py::test_duplicated_ignored_sequence_group, tests/test_regression.py::test_max_tokens_none, tests/test_regression.py::test_gc, tests/test_regression.py::test_model_from_modelscope
> ...
> [2024-11-21T18:53:51Z] INFO 11-21 10:53:51 model_runner.py:1399] Capturing cudagraphs for decoding. This may lead to unexpected consequences if the model is not static. To run the model in eager mode, set 'enforce_eager=True' or use '--enforce-eager' in the CLI.
> [2024-11-21T18:53:51Z] INFO 11-21 10:53:51 model_runner.py:1403] If out-of-memory error occurs during cudagraph capture, consider decreasing `gpu_memory_utilization` or switching to eager mode. You can also reduce the `max_num_seqs` as needed to decrease memory usage.
> [2024-11-21T18:54:02Z] INFO 11-21 10:54:02 model_runner.py:1517] Graph capturing finished in 10 secs, took 0.12 GiB
> Processed prompts:  50% 1/2 [00:00<00:00,  1.70it/s, est. speed input: 10.17 toks/s, output: 434.01 toks/s]
> # Received cancellation signal, interrupting
> [2024-11-21T20:52:13Z] 🚨 Error: The command exited with status -1
> ```

Whoops fixed that!

prashantgupta24 commented:
ugh this is a tough merge conflict

prashantgupta24 committed:
♻️ merge with main

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

prashantgupta24 committed:
Merge remote-tracking branch 'upstream/main' into prefill-slots

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

prashantgupta24 committed:
🎨 fmt

Signed-off-by: Prashant Gupta <prashantgupta@us.ibm.com>

prashantgupta24 commented:
@comaniac @njhill any chance we could get this merged? All tests passed

joerunde committed:
Merge remote-tracking branch 'upstream/main' into prefill-slots

joerunde commented on vllm/core/scheduler.py:
```diff
@@ -1755,6 +1935,8 @@ def _get_num_new_uncached_and_cached_tokens(
                 self._get_prompt_limit(seq_group),
                 num_uncached_new_tokens,
+                self.partial_prefill_budget_lookup_list,
+                partial_prefill_metadata,
```
nit: I think `_chunk_new_tokens_to_schedule` should not be static if this many of arguments are instance attributes

But doesn't have to block this PR

joerunde committed:
Merge remote-tracking branch 'upstream/main' into prefill-slots

rickyyx reviewed (commented):
This is definitely a great improvement for workload that might have long prefill requests. Also great job on the merge (I think I had a PR that also touched on very similar components)

W.r.t the PR itself, I mainly had one question: 
1. What's the overheads of recomputing partial prefill metadata on each scheduler step? If it's not significant, could we avoid it? I feel the information we needed for the partial metadata could be accumulated as some very limited amount of internal state of the scheduler. 

I am still a bit concerned of the potential consequences of having multiple concurrent prefills in the system, but if nothing breaks, then we are good. I will let folks with more knowledge on the other parts of the system to chime in. 

cc @comaniac

rickyyx commented on vllm/core/scheduler.py:
```diff
@@ -310,6 +321,98 @@ def scheduled_seq_group_builder():
+    phase faster.
+    """
+
+    # A minimum bound on the total number of prefills running during this
+    # scheduling step
+    partial_prefills: int
```
just a preliminary idea: do you think it would be possible to calculate the number of partial running prefills as we were doing the scheduling from the previous iteration and just track as part of scheduler's internal state?

joerunde replied:
> just a preliminary idea: do you think it would be possible to calculate the number of partial running prefills as we were doing the scheduling from the previous iteration and just track as part of scheduler's internal state?

Hah, that's what we were doing before but Cody suggested calculating and passing this metadata down to the scheduling methods where it's needed. So the goal here is to calculate just enough info at the beginning of each scheduler iteration to properly budget the partial prefills, hence the early breaks from looping through the queues.

> oh, we are also early breaking the loop, so this variable is kind of weird? it's then num_schedulable_prefills where shcedulable could be running and waiting?

Yeah that's an accurate name, maybe we can use that

joerunde committed:
Merge remote-tracking branch 'upstream/main' into prefill-slots

joerunde committed:
:bug: fix index out of range

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

joerunde commented on vllm/core/scheduler.py:
```diff
@@ -410,6 +514,18 @@ def __init__(
+        self.partial_prefill_budget_lookup_list = [0] * (
+            self.scheduler_config.max_num_partial_prefills + 1)
+        self.partial_prefill_budget_lookup_list[0] = (
+            scheduler_config.max_num_batched_tokens)
```
> partial_prefill_budget_lookup_list is a list where num partial prefills -> token budget for each partial prefill?

Yup! We calculate this once up-front when the scheduler is created so that we don't have to do a division for every sequence in every scheduling step. A list access measures slightly faster than an integer division, at least on the machines we have running in IBM cloud.

 > If we want to keep this, I think moving this into the metadata class might be better?

Yeah we could do that, I just wasn't quite sure how to make that properly static and cached on the metadata class, since we wouldn't want to re-create this list on every scheduler step

joerunde replied:
I spent a few minutes looking into this and I'm not sure there's a clean way to add a class attribute to a @dataclass, unfortunately. I'd rather not make the code even more hacky than this is!

What do you think between either:
1. Leaving as is, or
2. Doing the division in-line like `scheduler_config.max_num_batched_tokens // partial_prefill_metadata.schedulable_prefills` ?

joerunde replied:
Did some small benchmarking and running the division inline is noticeably slower, impacts e2e latency by like half a percent, so I'd like to leave as-is

joerunde commented on vllm/model_executor/layers/sampler.py:
```diff
@@ -995,7 +995,9 @@ def get_logprobs(
         empty_sampled_logprob: SampleLogprobs = []
         empty_prompt_logprob: Optional[PromptLogprobs] = None
-        return [empty_prompt_logprob], [empty_sampled_logprob]
+        num_seq_groups = len(sampling_metadata.seq_groups)
```
What we think is going on here is that the catches the case where for all sequences in the batch, no sequence generated a new token- i.e. all sequences are doing prefill and not yet decoding. The `return [empty_prompt_logprob], [empty_sampled_logprob]` statement seems to have been coded with the assumption that there would only ever be a batch size of 1 with all sequences prefilling. This makes the model return only a single output for any batch of prefill-only requests, which then causes things to break later. The change here just makes the output match the batch size

joerunde commented:
> 1. What's the overheads of recomputing partial prefill metadata on each scheduler step? If it's not significant, could we avoid it? I feel the information we needed for the partial metadata could be accumulated as some very limited amount of internal state of the scheduler.

@rickyyx good question! I kinda already answered why we went this way in [this comment](https://github.com/vllm-project/vllm/pull/10235#discussion_r1890785610) but another reason we need to do some amount of up-front processing is that we have to first peek through the waiting queue at the beginning of each step to see if there are prefills that we can schedule so that we can then properly budget the tokens when we schedule the running queue. So, we could keep info around in the scheduler state about the number of prefills currently running, but we still have to peek at the waiting queue each step.

As for exactly how long the `from_queues` method takes, I don't know. The results we measured from the serving benchmark suggest that there's no significant overhead, but if you want we could instrument this method with some timing and measure exactly how long it takes.

joerunde commented:
Related recent issue that I think would be improved by this: https://github.com/vllm-project/vllm/issues/11286

noooop commented:
maybe related to https://github.com/vllm-project/vllm/issues/10774

joerunde commented:
@rickyyx Here are some numbers from running the serving benchmark and measuring the time of `from_queues` 
```
Metadata Overhead (mean/median/p95/max): 0.005/0.005/0.010/0.027 ms
```

This took at most 27 nanoseconds, so I don't think we need to worry about this overhead. What do you think?

rickyyx commented:
> @rickyyx Here are some numbers from running the serving benchmark and measuring the time of `from_queues`
> 
> ```
> Metadata Overhead (mean/median/p95/max): 0.005/0.005/0.010/0.027 ms
> ```
> 
> This took at most 27 nanoseconds, so I don't think we need to worry about this overhead. What do you think?

Oh nice, this is rather negligible. Thanks for collecting that - do you know roughly how many elements are in the queue? Or what's the load when serving?

joerunde commented:
@rickyyx It was the "medium case" from the pr description at 12qps, but I didn't grab any logs from the server about the size of the queues while it was running. I can run the other cases and do it at higher qps to verify it doesn't get out of hand

joerunde commented:
@rickyyx Some more numbers, with the "mixed case" under 24qps load:

```
INFO 12-19 21:15:37 metrics.py:467] Avg prompt throughput: 6380.4 tokens/s, Avg generation throughput: 532.1 tokens/s, Running: 91 reqs, Swapped: 0 reqs, Pending: 230 reqs, GPU KV cache usage: 98.7%, CPU KV cache usage: 0.0%.
Metadata Overhead (mean/median/p95/max): 0.031/0.025/0.070/0.106 ms
```

With many requests running and a large waiting queue it maxes out at 100 nanoseconds. So it's a little worse, but I think still sub-millisecond should be fine here.

joerunde committed:
:recycle: naming updates

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

joerunde committed:
:bug: fix long prefill threshold init

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

joerunde commented on vllm/core/scheduler.py:
```diff
@@ -310,6 +321,98 @@ def scheduled_seq_group_builder():
+        waiting_partial_prefills = 0
+        waiting_long_prefills = 0
+
+        for sg in running:
```
(answered in comments below)

joerunde committed:
Merge remote-tracking branch 'upstream/main' into prefill-slots

joerunde commented:
Thanks @schoennenbeck!

Our teams are also now testing this after getting back from the holidays. I'm thinking I might move the configs to an environment variable like `VLLM_V0_MAX_PARTIAL_PREFILLS` since this change doesn't affect the incoming v1 scheduler and I don't want to cause confusion by creating new cli args that are immediately deprecated.

joerunde committed:
Merge remote-tracking branch 'upstream/main' into prefill-slots

joerunde commented:
Circling back on this: I've re-run the `Large Requests` benchmarks with this PR rebased on current main, and have also compared against using the V1 scheduler
![image](https://github.com/user-attachments/assets/3b6424af-9b75-45aa-baaa-c4095833092d)

The numbers have generally improved since this was first benchmarked, but this change still handily beats both main and V1 in TTFT and E2E latency. I might lean towards keeping these CLI args and adding this logic into the V1 scheduler- but of course will need more input on that as I know we're trying to keep V1 as simple as possible.

cc @comaniac @njhill, do y'all have thoughts on getting this in?

joerunde committed:
Merge remote-tracking branch 'upstream/main' into prefill-slots

joerunde committed:
Merge remote-tracking branch 'upstream/main' into prefill-slots

joerunde committed:
:art: fmt

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

njhill commented:
@comaniac WDYT about getting this PR in for v0? It makes a big difference in some of the IBM mixed production workloads, and per other comments above helps many others seeing similar problems.

We should probably extend our performance tests to incorporate these kinds of mixed workloads.

comaniac commented:
> @comaniac WDYT about getting this PR in for v0? It makes a big difference in some of the IBM mixed production workloads, and per other comments above helps many others seeing similar problems.
> 
> We should probably extend our performance tests to incorporate these kinds of mixed workloads.

No objection given the updated benchmark. v1 scheduler actually also supports multiple partial prefills just today (#12674), although it doesn't support configurations introduced by this PR. I'll take another pass tomorrow.

njhill commented:
> > @comaniac WDYT about getting this PR in for v0? It makes a big difference in some of the IBM mixed production workloads, and per other comments above helps many others seeing similar problems.
> > We should probably extend our performance tests to incorporate these kinds of mixed workloads.
> 
> No objection given the updated benchmark results. v1 scheduler actually also supports multiple partial prefills just today (#12674), although it doesn't support configurations introduced by this PR. I'll take another pass tomorrow.

Thanks @comaniac! Yes that's great re V1, I'm hoping that may allow us to incorporate similar logic there in a less invasive way.

comaniac commented on vllm/config.py:
```diff
@@ -1406,6 +1406,21 @@ class SchedulerConfig:
     # Maximum length of a sequence (including prompt and generated text).
     max_model_len: int = 8192
 
+    # Maximum number of sequences that can be partially prefilled concurrently
+    max_num_partial_prefills: int = 1
+
+    # Maximum number of "very long prompt" sequences that can be prefilled
+    # concurrently (long is defined by long_prefill_threshold)
+    max_long_partial_prefills: int = 1
+
+    # Set a percentage of the context length that determines which
+    # sequences are considered "long"
+    long_prefill_threshold: float = 0.04
+
+    # calculate context length that determines which sequences are
+    # considered "long"
+    long_prefill_token_threshold: int = 0
```
`long_prefill_token_threshold` is not used because you have the following?

```
        self.long_prefill_token_threshold = int(self.max_model_len *
                                                self.long_prefill_threshold)
```

I'd suggest to make the default behavior very clear and straightforward (but it doesn't have to be optimal). There are some example use cases I could think of:

1. If I know these configures after reading the document, and I want to properly set them to reduce TTFT. I firstly tried to increase `max_num_partial_prefills` and got stuck, because the differences between `max_num_partial_prefills` and `max_long_partial_prefills` are not clear.
2. `long_prefill_threshold` and its default value are not very straightforward, because 1) since users must be explicitly enabling this feature, we should enforce them to specify these values by themselves; 2) letting users specify the prefill token threshold seems easier in this case, as when you want to enable this feature, you must have already known the model length and the current performance, so you should also have a desired threshold in mind.

In addition, please validate these configurations in `_verify_args`. For example, it's not valid to enable them when chunked prefill is disabled, etc.

comaniac commented on vllm/core/scheduler.py:
```diff
@@ -294,23 +300,27 @@ def create_empty(cls) -> "SchedulerPrefillOutputs":
-                                 seq_data={},
-                                 sampling_params=None,
-                                 block_tables={})
+    return SequenceGroupMetadata(
+        request_id="",
+        is_prompt=False,
+        seq_data={},
+        sampling_params=None,
+        block_tables={},
+    )
```
Seems like formatting noise? If this was not done by the vLLM pre-commit, can you revert it to minimize the change in this file?

joerunde replied:
Yup, not sure how this got in but when it's reverted pre-commit doesn't complain

comaniac commented on vllm/core/scheduler.py:
```diff
@@ -319,6 +329,99 @@ def scheduled_seq_group_builder():
+
+    scheduler_config: SchedulerConfig
+
+    def cannot_schedule(self, seq_group: SequenceGroup) -> bool:
```
We already have `can_schedule` API in the scheduler. Can we follow that convention to use `can_schedule` here as well to make them consistent?

joerunde replied:
done!

comaniac commented on vllm/core/scheduler.py:
```diff
@@ -319,6 +329,99 @@ def scheduled_seq_group_builder():
+                >= self.scheduler_config.max_long_partial_prefills
+                and self.scheduler_config.max_num_partial_prefills > 1)
+
+    def increment_partial_prefills(self, seq_group: SequenceGroup) -> None:
```
It's not always incremented.

```suggestion
    def maybe_increment_partial_prefills(self, seq_group: SequenceGroup) -> None:
```

comaniac commented on vllm/core/scheduler.py:
```diff
@@ -932,10 +1069,19 @@ def _schedule_prefills(
+            if (partial_prefill_metadata is not None
+                    and partial_prefill_metadata.cannot_schedule(seq_group)):
+                leftover_waiting_sequences.appendleft(seq_group)
+                waiting_queue.popleft()
```
I feel it's safer to have `waiting_queue.popleft()` in the beginning (L1062) to make sure it is removed. Ideally `waiting_queue.popleft()` shouldn't appear more than one time in this function, but it now appears many times and I'm afraid that it may introduce bugs in the future if someone forget to pop it somewhere.

joerunde replied:
Yeah, I hear that but I also see the opposite problem: The waiting queue needs to _not_ be popped if we don't stick the sequence group in another queue (`leftover_waiting_sequences`, `ignored_sequence_groups`, or `running` in this case) otherwise we lose the sequence group completely.

It might be better to wrap these append and pops together in little methods to ensure that they're used together

comaniac commented on vllm/core/scheduler.py:
```diff
@@ -1810,29 +1995,31 @@ def _chunk_new_tokens_to_schedule(
+            # a sequence, we always allocate a number of new tokens that is
+            # divisible by the block size to avoid partial block matching.
             block_size = cache_config.block_size
-            remainder = budget.token_budget % block_size
-            if remainder != 0:
-                raise ValueError("When enabling chunked prefill and "
-                                 "prefix caching, max_num_batched_tokens "
-                                 "(chunk size) must be dividable by "
-                                 "block size, but got chunk_size "
-                                 f"({budget.token_budget}) % block_size "
-                                 f"({block_size}) = {remainder}")
```
Is this constraint relief?

comaniac reviewed (commented):
Otherwise LGTM

joerunde commented:
Thanks for the next pass @comaniac! I'll get on these as soon as I can- might be Monday though since I'm about to head out for a long weekend 🍹🍹🍹

joerunde committed:
Merge remote-tracking branch 'upstream/main' into prefill-slots

joerunde committed:
:wrench: update config and tests

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

joerunde committed:
:rewind: revert style change

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

joerunde committed:
:recycle: cannot -> can

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

joerunde committed:
Update vllm/core/scheduler.py

Signed-off-by: Joe Runde <joe@joerun.de>

Co-authored-by: Cody Yu <hao.yu.cody@gmail.com>

joerunde committed:
:art: fmt

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

joerunde committed:
:bug: fixup renamed fn ref

Signed-off-by: Joe Runde <Joseph.Runde@ibm.com>

comaniac reviewed (approved):
LGTM. We could merge once the CI is green

joerunde commented:
Thanks @comaniac!

I'll also run some final benchmarks to make sure it's still fast before merging

joerunde commented:
Quick final check on e2e latency on the "large sharegpt" benchmark:

| -- | 0.7.2 | This PR | This PR @4 concurrent prefills |
| -- | -- | -- | -- |
| median (ms) | 218 | 218 | 222 |
| p90 (ms) | 3683 | 3691 | 1576 |
| p95 (ms) | 5133 | 5144 | 1747 |
| p99 (ms) | 6449 | 6457 | 2522 |

njhill commented:
Thanks @joerunde @comaniac!!!

eladamittai commented:
Wow, this feature is very cool! Awesome PR! But the amount of long prefill requests concurrently processed is still limited but the budget yes? Let's say I have `--max-num-batched-tokens=256` and 130 running decode requests. I received 2 long prefill requests, let's say 3000 input tokens each. Will it divide the remaining budget between the 2 requests to execute them in parallel or will it just take the first request and chunk it as usual while the other one waits in the queue? And if it does divide it between the 2 requests, does it really improve the TTFT in this situation?

hibukipanim commented:
@joerunde there seem to be some typos regarding `--long-prefill-token-threshold`.

- the param description says "Defaults to 4%% of the model's context length.", but the default value is 0: https://github.com/vllm-project/vllm/blob/460f553a6d51c1ee10f8e86ec5829b473167b991/vllm/engine/arg_utils.py#L129
- in the ArgParser the type is `float` but in the `EngineArgs` dataclass it's `int`

and b.t.w it's not clear to me if it's 4% from the bottom or from the top. e.g. if the model has context-length of 100. 4% is context above 4 or above 96 ? as 4% sounds too small for being a threshold for "long", no?

schoennenbeck commented:
@hibukipanim Both 0 and None don't make sense as a threshold. The logic in the parser takes `int(.04 * model_max_len)` as the threshold if `long_prefill_token_threshold` is not set to a positive integer by the user.

It is indeed 4% from the bottom. With modern models this actually makes sense since they often have a max context length of ~100k and 4000 tokens is indeed already quite long (though for most use cases it probably still makes sense to tune this a little and not rely on the default).

madhavaggar commented:
With V1, I cannot set max_num_partial_prefills>1. Is this expected behavior, or is this a pending feature? What is the default behavior while enabling_chunked_prefill=True with vLLM V1? What settings can I use to optimize this for a large system prompt? And what is the best way to determine the best settings for these parameters? Is it by using the torch profiler and observing the time for each prefill chunk?
